### PR TITLE
write point cloud with custom attributes

### PR DESCRIPTION
**Pull request by ShubhamAgarwal12**
Write tensor point cloud with custom attributes support.

---
This change is [Reviewable](https://reviewable.io/reviews/intel-isl/open3d/2594)

### CODE REVIEW

File: cpp/open3d/t/io/file_format/FilePLY.cpp [hunk 1]
```diff
@@ -298,15 +299,183 @@ bool ReadPointCloudFromPLY(const std::string &filename,
```
Comment (ShubhamAgarwal12) [inline cpp/open3d/t/io/file_format/FilePLY.cpp:299]: To make sure the attributes have same format as in XYZI.

File: cpp/open3d/t/io/file_format/FilePLY.cpp [hunk 2]
```diff
@@ -298,15 +299,183 @@ bool ReadPointCloudFromPLY(const std::string &filename,
 
     // Add rest of the attributes.
     for (auto const &it : state.name_to_attr_state_) {
-        pointcloud.SetPointAttr(it.second->name_,
-                                core::TensorList::FromTensor(it.second->data_));
+        core::TensorList newAttr = core::TensorList(
+                it.second->size_, {1}, it.second->data_.GetDtype());
+        newAttr.AsTensor().IndexExtract(1, 0) = it.second->data_;
+        pointcloud.SetPointAttr(it.second->name_, newAttr);
```
Comment (ShubhamAgarwal12) [inline cpp/open3d/t/io/file_format/FilePLY.cpp:305]: To make sure the attributes have same format as in XYZI.

**Review by yxlao — CHANGES_REQUESTED** [review]
> Reviewed 4 of 5 files at r1, 1 of 1 files at r2.
> *[Reviewable](https://reviewable.io/reviews/intel-isl/open3d/2594)* status: all files reviewed, 14 unresolved discussions (waiting on @ShubhamAgarwal12)

___
*[cpp/open3d/t/geometry/PointCloud.h, line 121 at r2](https://reviewable.io/reviews/intel-isl/open3d/2594#-MLpfSjj6_6Cq7jhCuFH:-MLpfSjj6_6Cq7jhCuFI:biai964) ([raw file](https://github.com/intel-isl/open3d/blob/7cef74634a960d203fdd56035b48915270b76cc0/cpp/open3d/t/geometry/PointCloud.h#L121)):*
> ```objc
> 
>     /// Getter for point_attr_ TensorListMap. Used in Pybind.
>     const TensorListMap &GetPointAttrMap() const { return point_attr_; }
> ```

Does it work if we call it `GetPointAttr`?
___
*[cpp/open3d/t/io/file_format/FilePLY.cpp, line 305 at r1](https://reviewable.io/reviews/intel-isl/open3d/2594#-MLaokRe4NFmffAHWn9V-r1-305:-MLpigswDA5yrx_70DL_:bhzcgby) ([raw file](https://github.com/intel-isl/open3d/blob/6f3aa445aa158b5f69fefbd77f1c907535b96e7e/cpp/open3d/t/io/file_format/FilePLY.cpp#L305)):*
<details><summary><i>Previously, ShubhamAgarwal12 (Shubham Agarwal) wrote…</i></summary><blockquote>

To make sure the attributes have same format as in XYZI.
</blockquote></details>

could you elaborate on why we need this temporary `newAttr`?
___
*[cpp/open3d/t/io/file_format/FilePLY.cpp, line 313 at r2](https://reviewable.io/reviews/intel-isl/open3d/2594#-MLphnXO2NXSmpwcKTKd:-MLphnXO2NXSmpwcKTKe:b-nd1cnp) ([raw file](https://github.com/intel-isl/open3d/blob/7cef74634a960d203fdd56035b48915270b76cc0/cpp/open3d/t/io/file_format/FilePLY.cpp#L313)):*
> ```cpp
> }
> 
> static e_ply_type GetPlyType(core::Dtype d_type) {
> ```

dtype is not an enum, so pass with `const &`
there are some previous code that is pass-by-value, but this should be updated in the future
___
*[cpp/open3d/t/io/file_format/FilePLY.cpp, line 313 at r2](https://reviewable.io/reviews/intel-isl/open3d/2594#-MLpj5cF7tMlVgCmcJIO:-MLpj5cF7tMlVgCmcJIP:b7dtp2b) ([raw file](https://github.com/intel-isl/open3d/blob/7cef74634a960d203fdd56035b48915270b76cc0/cpp/open3d/t/io/file_format/FilePLY.cpp#L313)):*
> ```cpp
> }
> 
> static e_ply_type GetPlyType(core::Dtype d_type) {
> ```

`d_type` -> `dtype`
___
*[cpp/open3d/t/io/file_format/FilePLY.cpp, line 337 at r2](https://reviewable.io/reviews/intel-isl/open3d/2594#-MLpiv1V0Ez4jWkttgOE:-MLpiv1V0Ez4jWkttgOF:b7n439y) ([raw file](https://github.com/intel-isl/open3d/blob/7cef74634a960d203fdd56035b48915270b76cc0/cpp/open3d/t/io/file_format/FilePLY.cpp#L337)):*
> ```cpp
> template <typename T>
> static double GetValue(core::Tensor t_attr, int pos) {
>     T *data_ptr = static_cast<T *>(t_attr.GetDataPtr());
> ```

this can be `const T*` since we're not changing the ptr
___
*[cpp/open3d/t/io/file_format/FilePLY.cpp, line 349 at r2](https://reviewable.io/reviews/intel-isl/open3d/2594#-MLpl7MB44VYzEwCNzX2:-MLpl7MB44VYzEwCNzX3:b-qfcejy) ([raw file](https://github.com/intel-isl/open3d/blob/7cef74634a960d203fdd56035b48915270b76cc0/cpp/open3d/t/io/file_format/FilePLY.cpp#L349)):*
> ```cpp
>     }
> 
>     p_ply ply_file =
> ```

Move this after all checkings are done for the `pointcloud`. If `pointcloud` failed the check, we don't even create the file.
___
*[cpp/open3d/t/io/file_format/FilePLY.cpp, line 360 at r2](https://reviewable.io/reviews/intel-isl/open3d/2594#-MLplgWK1KpC5LkzF552:-MLplgWK1KpC5LkzF553:b-jy1x02) ([raw file](https://github.com/intel-isl/open3d/blob/7cef74634a960d203fdd56035b48915270b76cc0/cpp/open3d/t/io/file_format/FilePLY.cpp#L360)):*
> ```cpp
> 
>     geometry::TensorListMap tl_map = pointcloud.GetPointAttrMap();
>     long size = static_cast<long>(pointcloud.GetPoints().GetSize());
> ```

variable names should be informative, change this to `num_points`
___
*[cpp/open3d/t/io/file_format/FilePLY.cpp, line 363 at r2](https://reviewable.io/reviews/intel-isl/open3d/2594#-MLpnCaT8vS7BLsgYk7v:-MLpnCaT8vS7BLsgYk7w:b-yq8lwj) ([raw file](https://github.com/intel-isl/open3d/blob/7cef74634a960d203fdd56035b48915270b76cc0/cpp/open3d/t/io/file_format/FilePLY.cpp#L363)):*
> ```cpp
> 
>     // Make sure all the attribtes have same size.
>     for (auto const &it : tl_map) {
> ```

use `TensorListMap::IsSizeSynchronized()`
___
*[cpp/open3d/t/io/file_format/FilePLY.cpp, line 412 at r2](https://reviewable.io/reviews/intel-isl/open3d/2594#-MLpmGsa-9m8OpdvmknD:-MLpmGsa-9m8OpdvmknE:b3jdjca) ([raw file](https://github.com/intel-isl/open3d/blob/7cef74634a960d203fdd56035b48915270b76cc0/cpp/open3d/t/io/file_format/FilePLY.cpp#L412)):*
> ```cpp
>         }
>     }
> ```

it would be more elegant to group the several for loops into one:

```cpp
for (const auto & it: tl_map) {
    if (it.fisrt == "points") {
        // add x, y, z
    } else if (it.first == "normals") {
        // add nx, ny, nz
    } else if (it.first == "colors") {
        // ...
    } else {
       // ....
    }
}
```

This way, you can avoid calling `pointcloud.HasPointNormals()` and etc.
___
*[cpp/open3d/t/io/file_format/FilePLY.cpp, line 420 at r2](https://reviewable.io/reviews/intel-isl/open3d/2594#-MLpnNRe17ns02q82_Jt:-MLpnNRe17ns02q82_Ju:b-tcn5r) ([raw file](https://github.com/intel-isl/open3d/blob/7cef74634a960d203fdd56035b48915270b76cc0/cpp/open3d/t/io/file_format/FilePLY.cpp#L420)):*
> ```cpp
> 
>     utility::CountingProgressReporter reporter(params.update_progress);
>     reporter.SetTotal(pointcloud.GetPoints().GetSize());
> ```

we've already computed `num_points` before, use that
___
*[cpp/open3d/t/io/file_format/FilePLY.cpp, line 422 at r2](https://reviewable.io/reviews/intel-isl/open3d/2594#-MLpoQ5y9ae7M6aDFSAg:-MLpnRZCD7a7LWVleTqa:b-btmbvo) ([raw file](https://github.com/intel-isl/open3d/blob/7cef74634a960d203fdd56035b48915270b76cc0/cpp/open3d/t/io/file_format/FilePLY.cpp#L422)):*
> ```cpp
>     reporter.SetTotal(pointcloud.GetPoints().GetSize());
> 
>     for (int64_t i = 0; i < pointcloud.GetPoints().GetSize(); i++) {
> ```

same, use `num_points`
___
*[cpp/open3d/t/io/file_format/FilePLY.cpp, line 425 at r2](https://reviewable.io/reviews/intel-isl/open3d/2594#-MLpywwP3TNQWgiKuTN8:-MLpywwQ-7UM1DJX7pDq:b-zambnp) ([raw file](https://github.com/intel-isl/open3d/blob/7cef74634a960d203fdd56035b48915270b76cc0/cpp/open3d/t/io/file_format/FilePLY.cpp#L425)):*
> ```cpp
>         DISPATCH_DTYPE_TO_TEMPLATE(pointcloud.GetPoints().GetDtype(), [&]() {
>             ply_write(ply_file,
>                       GetValue<scalar_t>(pointcloud.GetPoints()[i], 0));
> ```

1. `pointcloud.GetPoints()[i]` still triggers tensor indexing, thus, we can to avoid the `[i]` by computing offset like `ptr + i * 3 + j`
2. `GetPoints()` is called 3 times per point, we can avoid this by `offset = ptr + i * 3;` and use `offset + 0, offset + 1, offset + 2`
___
*[cpp/open3d/t/io/file_format/FilePLY.cpp, line 460 at r2](https://reviewable.io/reviews/intel-isl/open3d/2594#-MLpoiBACgdhOUqeuj6_:-MLpoiBACgdhOUqeuj6a:b-13m9yr) ([raw file](https://github.com/intel-isl/open3d/blob/7cef74634a960d203fdd56035b48915270b76cc0/cpp/open3d/t/io/file_format/FilePLY.cpp#L460)):*
> ```cpp
>         }
> 
>         for (auto const &it : tl_map) {
> ```

same comment as before, group the conditions into one

```
if (is points) {
     xxx
} else if (is colors) {
     xxx
} else if (is normals) {
     xxx
} else {
     xxx
}
```
___
*[cpp/open3d/t/io/file_format/FilePLY.cpp, line 477 at r2](https://reviewable.io/reviews/intel-isl/open3d/2594#-MLplO2_Ccb4ga9zLs7_:-MLplO2_Ccb4ga9zLs7a:b-ifcr1h) ([raw file](https://github.com/intel-isl/open3d/blob/7cef74634a960d203fdd56035b48915270b76cc0/cpp/open3d/t/io/file_format/FilePLY.cpp#L477)):*
> ```cpp
>     ply_close(ply_file);
>     return true;
> }
> ```

1. Add unit test (before addressing comments, so that we make sure the correctness). 
2. Can you also provide a python benchmark comparison script as github comments and report benchmark results comparing this with the original writer?

**Review by ShubhamAgarwal12 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/intel-isl/open3d/2594)* status: 4 of 6 files reviewed, 14 unresolved discussions (waiting on @yxlao)

___
*[cpp/open3d/t/geometry/PointCloud.h, line 121 at r2](https://reviewable.io/reviews/intel-isl/open3d/2594#-MLpfSjj6_6Cq7jhCuFH:-MLv-v_DBuNK103e2RdG:bvz3ve1) ([raw file](https://github.com/intel-isl/open3d/blob/7cef74634a960d203fdd56035b48915270b76cc0/cpp/open3d/t/geometry/PointCloud.h#L121)):*
<details><summary><i>Previously, yxlao (Yixing Lao) wrote…</i></summary><blockquote>

Does it work if we call it `GetPointAttr`?
</blockquote></details>

It works fine in the c++ side but in python api, it gives error. There is another function with same name but with the attribute key as argument. 

TypeError: (): incompatible function arguments. The following argument types are supported:
    1. (arg0: open3d.cpu.pybind.t.geometry.PointCloud, arg1: str) -> open3d.cpu.pybind.core.TensorList

Invoked with: <open3d.cpu.pybind.t.geometry.PointCloud object at 0x109b3adb0>
___
*[cpp/open3d/t/io/file_format/FilePLY.cpp, line 305 at r1](https://reviewable.io/reviews/intel-isl/open3d/2594#-MLaokRe4NFmffAHWn9V-r1-305:-MLv1poI4HhWSkMqzcZr:bwrmdmk) ([raw file](https://github.com/intel-isl/open3d/blob/6f3aa445aa158b5f69fefbd77f1c907535b96e7e/cpp/open3d/t/io/file_format/FilePLY.cpp#L305)):*
<details><summary><i>Previously, yxlao (Yixing Lao) wrote…</i></summary><blockquote>

could you elaborate on why we need this temporary `newAttr`?
</blockquote></details>

In XYZI, the Attribute TensorLists have element_shape (1) whereas for PLY we directly convert the tensor to tensorList and its element_shape is 0-sized. I thought, we should have uniformity in this implementation. That also allows us to have common unit tests. The "element_shape" ref is given below.
 (PS: I added the tests, but the unit test addition got missed in the previous commit. Did it now.)


TensorList(int64_t size, const open3d::core::SizeVector &element_shape, open3d::core::Dtype dtype, const open3d::core::Device &device = ((open3d::core::Device)("CPU:0")))
___
*[cpp/open3d/t/io/file_format/FilePLY.cpp, line 313 at r2](https://reviewable.io/reviews/intel-isl/open3d/2594#-MLphnXO2NXSmpwcKTKd:-MLv52TY6pOxaUJ8Ow-P:b-896fix) ([raw file](https://github.com/intel-isl/open3d/blob/7cef74634a960d203fdd56035b48915270b76cc0/cpp/open3d/t/io/file_format/FilePLY.cpp#L313)):*
<details><summary><i>Previously, yxlao (Yixing Lao) wrote…</i></summary><blockquote>

dtype is not an enum, so pass with `const &`
there are some previous code that is pass-by-value, but this should be updated in the future
</blockquote></details>

Done.
___
*[cpp/open3d/t/io/file_format/FilePLY.cpp, line 313 at r2](https://reviewable.io/reviews/intel-isl/open3d/2594#-MLpj5cF7tMlVgCmcJIO:-MLv534t-pBqxrqhj_RP:b-896fix) ([raw file](https://github.com/intel-isl/open3d/blob/7cef74634a960d203fdd56035b48915270b76cc0/cpp/open3d/t/io/file_format/FilePLY.cpp#L313)):*
<details><summary><i>Previously, yxlao (Yixing Lao) wrote…</i></summary><blockquote>

`d_type` -> `dtype`
</blockquote></details>

Done.
___
*[cpp/open3d/t/io/file_format/FilePLY.cpp, line 337 at r2](https://reviewable.io/reviews/intel-isl/open3d/2594#-MLpiv1V0Ez4jWkttgOE:-MLv5VL2-OkKC3vf1rFK:b-896fix) ([raw file](https://github.com/intel-isl/open3d/blob/7cef74634a960d203fdd56035b48915270b76cc0/cpp/open3d/t/io/file_format/FilePLY.cpp#L337)):*
<details><summary><i>Previously, yxlao (Yixing Lao) wrote…</i></summary><blockquote>

this can be `const T*` since we're not changing the ptr
</blockquote></details>

Done.
___
*[cpp/open3d/t/io/file_format/FilePLY.cpp, line 349 at r2](https://reviewable.io/reviews/intel-isl/open3d/2594#-MLpl7MB44VYzEwCNzX2:-MLv5zs3BeEkPhC6EwH2:b-896fix) ([raw file](https://github.com/intel-isl/open3d/blob/7cef74634a960d203fdd56035b48915270b76cc0/cpp/open3d/t/io/file_format/FilePLY.cpp#L349)):*
<details><summary><i>Previously, yxlao (Yixing Lao) wrote…</i></summary><blockquote>

Move this after all checkings are done for the `pointcloud`. If `pointcloud` failed the check, we don't even create the file.
</blockquote></details>

Done.
___
*[cpp/open3d/t/io/file_format/FilePLY.cpp, line 360 at r2](https://reviewable.io/reviews/intel-isl/open3d/2594#-MLplgWK1KpC5LkzF552:-MLv7F8p2_l1RDik_IDR:b-896fix) ([raw file](https://github.com/intel-isl/open3d/blob/7cef74634a960d203fdd56035b48915270b76cc0/cpp/open3d/t/io/file_format/FilePLY.cpp#L360)):*
<details><summary><i>Previously, yxlao (Yixing Lao) wrote…</i></summary><blockquote>

variable names should be informative, change this to `num_points`
</blockquote></details>

Done.
___
*[cpp/open3d/t/io/file_format/FilePLY.cpp, line 363 at r2](https://reviewable.io/reviews/intel-isl/open3d/2594#-MLpnCaT8vS7BLsgYk7v:-MLv7Fmb9V_FxxEi4ABn:b-896fix) ([raw file](https://github.com/intel-isl/open3d/blob/7cef74634a960d203fdd56035b48915270b76cc0/cpp/open3d/t/io/file_format/FilePLY.cpp#L363)):*
<details><summary><i>Previously, yxlao (Yixing Lao) wrote…</i></summary><blockquote>

use `TensorListMap::IsSizeSynchronized()`
</blockquote></details>

Done.
___
*[cpp/open3d/t/io/file_format/FilePLY.cpp, line 412 at r2](https://reviewable.io/reviews/intel-isl/open3d/2594#-MLpmGsa-9m8OpdvmknD:-MM-9iJwDrXon1gZhDDo:b-i3f7at) ([raw file](https://github.com/intel-isl/open3d/blob/7cef74634a960d203fdd56035b48915270b76cc0/cpp/open3d/t/io/file_format/FilePLY.cpp#L412)):*
<details><summary><i>Previously, yxlao (Yixing Lao) wrote…</i></summary><blockquote>

it would be more elegant to group the several for loops into one:

```cpp
for (const auto & it: tl_map) {
    if (it.fisrt == "points") {
        // add x, y, z
    } else if (it.first == "normals") {
        // add nx, ny, nz
    } else if (it.first == "colors") {
        // ...
    } else {
       // ....
    }
}
```

This way, you can avoid calling `pointcloud.HasPointNormals()` and etc.
</blockquote></details>

If we club the conditions, the order for the attributes might not be standard as unordered map can have the elements in any order. For the PLY normally the base attributes are always added first.
___
*[cpp/open3d/t/io/file_format/FilePLY.cpp, line 420 at r2](https://reviewable.io/reviews/intel-isl/open3d/2594#-MLpnNRe17ns02q82_Jt:-MLv7Y4o5or-VCdb4fyz:b-896fix) ([raw file](https://github.com/intel-isl/open3d/blob/7cef74634a960d203fdd56035b48915270b76cc0/cpp/open3d/t/io/file_format/FilePLY.cpp#L420)):*
<details><summary><i>Previously, yxlao (Yixing Lao) wrote…</i></summary><blockquote>

we've already computed `num_points` before, use that
</blockquote></details>

Done.
___
*[cpp/open3d/t/io/file_format/FilePLY.cpp, line 422 at r2](https://reviewable.io/reviews/intel-isl/open3d/2594#-MLpoQ5y9ae7M6aDFSAg:-MLvDO_I2_Kfph3O_DmE:b-896fix) ([raw file](https://github.com/intel-isl/open3d/blob/7cef74634a960d203fdd56035b48915270b76cc0/cpp/open3d/t/io/file_format/FilePLY.cpp#L422)):*
<details><summary><i>Previously, yxlao (Yixing Lao) wrote…</i></summary><blockquote>

same, use `num_points`
</blockquote></details>

Done.
___
*[cpp/open3d/t/io/file_format/FilePLY.cpp, line 425 at r2](https://reviewable.io/reviews/intel-isl/open3d/2594#-MLpywwP3TNQWgiKuTN8:-MLvQtRm9KemDtpNecLX:b-9ahj9v) ([raw file](https://github.com/intel-isl/open3d/blob/7cef74634a960d203fdd56035b48915270b76cc0/cpp/open3d/t/io/file_format/FilePLY.cpp#L425)):*
<details><summary><i>Previously, yxlao (Yixing Lao) wrote…</i></summary><blockquote>

1. `pointcloud.GetPoints()[i]` still triggers tensor indexing, thus, we can to avoid the `[i]` by computing offset like `ptr + i * 3 + j`
2. `GetPoints()` is called 3 times per point, we can avoid this by `offset = ptr + i * 3;` and use `offset + 0, offset + 1, offset + 2`
</blockquote></details>

I tried the first one earlier. For me, using indexing was faster than using without the indexing in case of 1.  I thought, this is because we utilize std::vectors and that is already optimised for indexing.

I've done the change suggested in 2. 
___
*[cpp/open3d/t/io/file_format/FilePLY.cpp, line 460 at r2](https://reviewable.io/reviews/intel-isl/open3d/2594#-MLpoiBACgdhOUqeuj6_:-MM-A7Tc6uKTGpIygKY7:b-co4g7f) ([raw file](https://github.com/intel-isl/open3d/blob/7cef74634a960d203fdd56035b48915270b76cc0/cpp/open3d/t/io/file_format/FilePLY.cpp#L460)):*
<details><summary><i>Previously, yxlao (Yixing Lao) wrote…</i></summary><blockquote>

same comment as before, group the conditions into one

```
if (is points) {
     xxx
} else if (is colors) {
     xxx
} else if (is normals) {
     xxx
} else {
     xxx
}
```
</blockquote></details>

Same as previous. This might mess up the order.
___
*[cpp/open3d/t/io/file_format/FilePLY.cpp, line 477 at r2](https://reviewable.io/reviews/intel-isl/open3d/2594#-MLplO2_Ccb4ga9zLs7_:-MM1IPaU2bK9uxfOxQ4Y:bs1stdm) ([raw file](https://github.com/intel-isl/open3d/blob/7cef74634a960d203fdd56035b48915270b76cc0/cpp/open3d/t/io/file_format/FilePLY.cpp#L477)):*
<details><summary><i>Previously, yxlao (Yixing Lao) wrote…</i></summary><blockquote>

1. Add unit test (before addressing comments, so that we make sure the correctness). 
2. Can you also provide a python benchmark comparison script as github comments and report benchmark results comparing this with the original writer?
</blockquote></details>

```
import open3d as o3d
import time
import numpy as np

if __name__ == "__main__":
    file_name = "<Path_to_Open3D>/Open3D/examples/test_data/fragment.ply"
    iterations = 10

    # Benchmark T IO
    pct = o3d.t.io.read_point_cloud(file_name)
    s = time.time()
    for i in range(iterations):
        o3d.t.io.write_point_cloud("test_new.ply",pct, True, True)
    print("Time for t.io: ", (time.time() - s) / iterations)
    pctn = o3d.t.io.read_point_cloud("test_new.ply")

    # Benchmark old IO
    pcd = o3d.io.read_point_cloud(file_name)
    s = time.time()
    for i in range(iterations):
        o3d.io.write_point_cloud("test_old.ply",pcd, True, True)
    print("Time for io: ", (time.time() - s) / iterations)
    pcdn = o3d.io.read_point_cloud("test_old.ply")

    # Sanity check
    print(pctn.point["points"])
    np.testing.assert_allclose(pctn.point["points"].as_tensor().numpy(),
                               np.asarray(pcdn.points))
```

**Review by yxlao — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/intel-isl/open3d/2594)* status: 4 of 6 files reviewed, 14 unresolved discussions (waiting on @ShubhamAgarwal12 and @yxlao)

___
*[cpp/open3d/t/geometry/PointCloud.h, line 121 at r2](https://reviewable.io/reviews/intel-isl/open3d/2594#-MLpfSjj6_6Cq7jhCuFH:-MMTwKRu2igWXXEZlL9A:bglyr74) ([raw file](https://github.com/intel-isl/open3d/blob/7cef74634a960d203fdd56035b48915270b76cc0/cpp/open3d/t/geometry/PointCloud.h#L121)):*
<details><summary><i>Previously, ShubhamAgarwal12 (Shubham Agarwal) wrote…</i></summary><blockquote>

It works fine in the c++ side but in python api, it gives error. There is another function with same name but with the attribute key as argument. 

TypeError: (): incompatible function arguments. The following argument types are supported:
    1. (arg0: open3d.cpu.pybind.t.geometry.PointCloud, arg1: str) -> open3d.cpu.pybind.core.TensorList

Invoked with: <open3d.cpu.pybind.t.geometry.PointCloud object at 0x109b3adb0>



</blockquote></details>

This can be resolved by update the pybind code: https://pybind11.readthedocs.io/en/stable/classes.html#overloaded-methods
___
*[cpp/open3d/t/io/file_format/FilePLY.cpp, line 305 at r1](https://reviewable.io/reviews/intel-isl/open3d/2594#-MLaokRe4NFmffAHWn9V-r1-305:-MMTy6nb6WOOMuAHfYft:b-mtbvve) ([raw file](https://github.com/intel-isl/open3d/blob/6f3aa445aa158b5f69fefbd77f1c907535b96e7e/cpp/open3d/t/io/file_format/FilePLY.cpp#L305)):*
<details><summary><i>Previously, ShubhamAgarwal12 (Shubham Agarwal) wrote…</i></summary><blockquote>

In XYZI, the Attribute TensorLists have element_shape (1) whereas for PLY we directly convert the tensor to tensorList and its element_shape is 0-sized. I thought, we should have uniformity in this implementation. That also allows us to have common unit tests. The "element_shape" ref is given below.
 (PS: I added the tests, but the unit test addition got missed in the previous commit. Did it now.)


TensorList(int64_t size, const open3d::core::SizeVector &element_shape, open3d::core::Dtype dtype, const open3d::core::Device &device = ((open3d::core::Device)("CPU:0")))
</blockquote></details>

Depending on the understanding, 0-sized `element_shape` can be considered as more consistent, since we're always taking away the first dimension, so a 0-sized "second" dimension is left.

Can you point to the specific location where 0-sized `element_shape`causes inconsistency?
___
*[cpp/open3d/t/io/file_format/FilePLY.cpp, line 337 at r2](https://reviewable.io/reviews/intel-isl/open3d/2594#-MLpiv1V0Ez4jWkttgOE:-MMTyy3SDkdnedUb0qTa:b-kjkqf2) ([raw file](https://github.com/intel-isl/open3d/blob/7cef74634a960d203fdd56035b48915270b76cc0/cpp/open3d/t/io/file_format/FilePLY.cpp#L337)):*
<details><summary><i>Previously, ShubhamAgarwal12 (Shubham Agarwal) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

what i mean is `static_cast<const T*>` and we still `static_cast` to `double` value since that's what we need. This can avoid casting in other locations.

**Review by ShubhamAgarwal12 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/intel-isl/open3d/2594)* status: 2 of 6 files reviewed, 13 unresolved discussions (waiting on @yxlao)

___
*[cpp/open3d/t/geometry/PointCloud.h, line 121 at r2](https://reviewable.io/reviews/intel-isl/open3d/2594#-MLpfSjj6_6Cq7jhCuFH:-MMYx38V5JTvVmm-kfM7:b-896fix) ([raw file](https://github.com/intel-isl/open3d/blob/7cef74634a960d203fdd56035b48915270b76cc0/cpp/open3d/t/geometry/PointCloud.h#L121)):*
<details><summary><i>Previously, yxlao (Yixing Lao) wrote…</i></summary><blockquote>

This can be resolved by update the pybind code: https://pybind11.readthedocs.io/en/stable/classes.html#overloaded-methods
</blockquote></details>

Done.
___
*[cpp/open3d/t/io/file_format/FilePLY.cpp, line 337 at r2](https://reviewable.io/reviews/intel-isl/open3d/2594#-MLpiv1V0Ez4jWkttgOE:-MMZ5kx-EjXAW6paR1aJ:bd5wowq) ([raw file](https://github.com/intel-isl/open3d/blob/7cef74634a960d203fdd56035b48915270b76cc0/cpp/open3d/t/io/file_format/FilePLY.cpp#L337)):*
<details><summary><i>Previously, yxlao (Yixing Lao) wrote…</i></summary><blockquote>

what i mean is `static_cast<const T*>` and we still `static_cast` to `double` value since that's what we need. This can avoid casting in other locations.
</blockquote></details>

I've made the change for const, but for return I still use  T instead of double. The problem is that we return the whole Tensor instead of individual value for the base attributes and access the individual values in the WritePointCloudToPLY function.

**Review by yxlao — APPROVED** [review]
> Reviewed 3 of 4 files at r3, 1 of 1 files at r4.
> *[Reviewable](https://reviewable.io/reviews/intel-isl/open3d/2594)* status: all files reviewed, 13 unresolved discussions (waiting on @yxlao)

**Pull request merged**